docs(cookbooks): adding creating a behat extension (#146)
Behat/docs: cookbooks/creating_a_behat_extension.rst

Creating a Behat extension
==========================

Extensions are particularly useful when configuration becomes a necessity.

In this cookbook, we will create a simple extension named ``LogsExtension`` that logs scenario durations and cover:

#. Setting up the context
#. Creating the extension
#. Initializing the context

Setting Up the Context
----------------------

> First, we need to create a context that will handle the logging.

Reviewer: I wouldn't necessarily say you *need* to implement a context if you only want to listen to events. It's not wrong,
and it works well in your example, but there are other solutions that might be better sometimes.

Maybe something like this?

```suggestion
First, we need to create a class that will handle the logging. There are several ways to do this (you can hook
directly into behat's event dispatcher, for example).

For our case, we will create a `Context` and use the :doc:`normal tagged hooks</user_guide/context/hooks>`
that you might have used in your own contexts.

This is a straightforward approach, and means we can also show how to initialise any context with custom
configuration or dependencies.
```

      public function before(BeforeScenarioScope $scope)
      {
          if (/* enable config */ === false) {
              return;
          }

          file_put_contents(
              /* filepath */,
              'START: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
              FILE_APPEND
          );
      }
  
      /** @AfterScenario */
      public function after(AfterScenarioScope $scope)
      {
          if (/* enable config */ === false) {
              return;
          }
  
          file_put_contents(
              /* filepath */,
              'END: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
              FILE_APPEND
          );
      }
  }

Creating the Extension
----------------------

Next, we need to create the extension itself.
This will serve as the entry point for our logging functionality.

Directory structure:

.. code-block::
  
  src/
      Context/
          LogsContext.php
      ServiceContainer/
          LogsExtension.php   # This is where we'll define our extension

To ensure Behat can find and load the ``LogsExtension.php`` file, it is important to place it within the `ServiceContainer` folder.

> While there might be alternatives, we will stick to the straightforward method.

Reviewer: I'm not certain this is correct - I think it can be called & located anywhere that your composer autoloader can find it?

Author: This is the part where I was stuck, not understanding why my extension wouldn't load _(kudos @TimoBakx for helping me on Symfony's Slack 🙏)_.

From my research, extensions seems to be loaded via the [ExtensionManager](https://github.com/Behat/Behat/blob/42f15e5d513fc10dd807f90379652bacd56930b6/src/Behat/Testwork/ServiceContainer/ExtensionManager.php#L176-L198). `$locator` is the key value used to define the extension configuration in the `behat.yml` file.

There four ways for the manager to load the extension:
```php
        if (class_exists($class = $locator)) {
            return new $class;
        }

        if (class_exists($class = $this->getFullExtensionClass($locator))) {
            return new $class;
        }

        if (file_exists($locator)) {
            return require($locator);
        }

        if (file_exists($path = $this->extensionsPath . DIRECTORY_SEPARATOR . $locator)) {
            return require($path);
        }
```

But I didn't know why, it was not working on putting the full namespace of the class wasn't enough in my tests, I had to put it in the `ServiceContainer` folder.

So within the post I choose to go the easy way, but we can change it.
I can try to understand why it wasn't working at the first place.

Reviewer: @monitaurus using the fully-qualified class for the extension does work and is applied before any of the subsequent magic in that method. If that wasn't working for you my best guess would be your class was in a file that composer couldn't autoload (so the `class_exists($locator)` at the top of that quoted code block would have returned false).

Author: I tried again, with the new `HelloWorld` extension, but it didn't work.
I'll try some step debug to find out, but I just let the full class namespace in the cookbook for simplicity.

      }
  }

.. note::
  
  The ``initialize`` and ``process`` methods are empty in our case but are useful when you need to interact with other extensions or process the container after it has been compiled.

Initializing the Context
------------------------

To pass configuration values to our ``LogsContext``, we need to create an initializer.

Directory structure:

.. code-block::
  src/
      Context/
          Initializer/
              LogsInitializer.php   # This will handle context initialization
          LogsContext.php
      ServiceContainer/
          LogsExtension.php

The code for ``LogsInitializer.php``:

.. code-block:: php  
  <?php

  namespace Behat\LogsExtension\Context\Initializer;

  use Behat\LogsExtension\Context\LogsContext;
  use Behat\Behat\Context\Context;
  use Behat\Behat\Context\Initializer\ContextInitializer;

  class LogsInitializer implements ContextInitializer
  {
      private string $filepath;
      private bool $enable;
  
      public function __construct(string $filepath, bool $enable)
      {
          $this->filepath = $filepath;
          $this->enable = $enable;
      }
  
      public function initializeContext(Context $context)
      {

>           if (!$context instanceof LogsContext) {
>               return;
>           }

Reviewer: Might be useful to explain a little more what this method is? It was the "aha" moment for me when I started trying to write / use extensions...

```suggestion
          /* 
           * At the start of every scenario, behat will create a new instance of every `Context`
           * registered in your project. It will then call this method with each new `Context` in
           * turn. If you want to initialise multiple contexts, you can of course give them an 
           * interface and check for that here.
           */
          if (!$context instanceof LogsContext) {
              return;
          }
```

      // ...
  }

To complete the extension, we must add methods to ``LogsContext`` to receive the configuration values and use those in the hooks:

.. code-block:: php

  // ...
  
  class LogsContext implements Context
  {
      private bool $enable = false;
      private string $filepath;
  
      public function initializeConfig(bool $enable, string $filepath)
      {
          $this->enable = $enable;
          $this->filepath = $filepath;
      }
  
      /** @BeforeScenario */
      public function before(BeforeScenarioScope $scope)
      {
          if ($this->enable === false) {
              return;
          }

          file_put_contents(
              $this->filepath,
              'START: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
              FILE_APPEND
          );
      }
  
      /** @AfterScenario */
      public function after(AfterScenarioScope $scope)
      {
          if ($this->enable === false) {
              return;
          }
  
          file_put_contents(
              $this->filepath,
              'END: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
              FILE_APPEND
          );
      }
  }

> Conclusion

Reviewer: The only final thing would be a quick example of what you need in your project's `behat.yml` to use this extension - I think in this case you'll need to register both the extension and the LogsContext separately so it would be worth showing that?

Author: Indeed it would be far better, I'll push a new commit soon with that.

And yes, it needs to register the context in the suite, and the extension in the profile: https://github.com/monitaurus/behat-extension-boilerplate/blob/main/behat.yml
The file is from a boilerplate repo I've made alongside my post to allow people to test it

Author: @acoulton done with f91f7c9

----------

Congratulations! You have just created a simple Behat extension that logs scenario durations. This extension demonstrates the three essential steps to building a Behat extension: defining an extension, creating an initializer, and configuring contexts.

> Feel free to experiment with this extension and expand its functionality. For further learning, check out the `Behat hooks documentation <https://behat.org/en/latest/user_guide/context/hooks.html>`_ and explore existing extensions on `GitHub <https://github.com/search?o=desc&q=behat+extension+in%3Aname%2Cdescription+language%3APHP&ref=searchresults&s=stars&type=Repositories>`_.

Reviewer: Just updated this to be an internal link within the docs engine.

```suggestion
Feel free to experiment with this extension and expand its functionality. For further learning, check out the :doc:`Behat hooks documentation</user_guide/context/hooks>` and explore existing extensions on `GitHub <https://github.com/search?o=desc&q=behat+extension+in%3Aname%2Cdescription+language%3APHP&ref=searchresults&s=stars&type=Repositories>`_.
```

Author: Thanks, I must admit I'm not at ease with rst format 😅

Reviewer: Me either :laughing: - thanks!

> Creating a Behat extension

Reviewer: Now this is specific to initialising contexts, I agree with the suggestion to rename it.

Maybe call the file `creating_a_context_configuration_extension.rst` (and update the link in cookbooks.rst) and then change the title to something like:

```suggestion
Creating a Behat extension to provide custom configuration for Contexts
```

Author: Applied ✅  and file renamed

==========================

Extensions are particularly useful when configuration becomes a necessity.

In this cookbook, we will create a simple extension named ``HelloWorld`` that will display some text and cover:

#. Setting up the context
#. Creating the extension
#. Initializing the context
#. Using the extension

Setting Up the Context
----------------------

First, we need to create a ``Context`` class that will throw a ``PendingException`` with a configurable text.

> A behavior that could be enable or not.

Reviewer: ```suggestion
The configuration will also control whether the behaviour is enabled or not.
```

  use Behat\Behat\Context\ServiceContainer\ContextExtension;

  class HelloWorldExtension implements Extension
  {
      // ...

      public function load(ContainerBuilder $container, array $config)
      {
          $definition = new Definition(HelloWorldInitializer::class, [
              $config['text'],
              $config['enable'],
          ]);
          $definition->addTag(ContextExtension::INITIALIZER_TAG);
          $container->setDefinition('helloworld_extension.context_initializer', $definition);
      }

      // ...
  }

To complete the extension, we must add methods to ``HelloWorldContext`` to receive the configuration values and use those in the hooks:

.. code-block:: php

  // ...

  class HelloWorldContext implements Context
  {
      private bool $enable = false;
      private string $text;

      public function initializeConfig(bool $enable, string $text)
      {
          $this->enable = $enable;
          $this->text = $text;
      }

      /** @Given I say Hello World */
      public function helloWorld()
      {
          if ($this->enable === false) {
              return;
          }
  
          throw new PendingException($this->text);
      }

>   }

Reviewer: I don't think this code block is functionally different to the first one (just the `helloWorld` method is formatted/structured slightly different. Did you mean to leave out the `initializeConfig` method in the first example and then add it here? 

I think the example is simple enough now that you could probably just include the whole `HelloWorldContext` once at the beginning and explain that the main job of the extension is to call `initializeConfig` with the values? And then leave this section out?

Author: It came from a copy / paste from a test repo, and forgot to remove the method.

But you're right, it simple enough now to have a single block at the beginning.
It's done 👌

Using the extension
-------------------

Now that the extension is ready and will inject values into context, we just need to configure it into a project.

In the ``extensions`` key of a profile (``default`` in our case), we'll add the ``HelloWorldExtension`` key and configure our ``text`` and ``enable`` value.

Finally, we need to load the ``HelloWorld\Context\HelloWorldContext`` into our suite.

Here's the ``behat.yaml``:

.. code-block:: yaml

  default:
    suites:
      default:
        contexts:
          - FeatureContext
          - HelloWorld\Context\HelloWorldContext
    extensions:
      HelloWorld\ServiceContainer\HelloWorldExtension:
        text: 'Hi there!'
        enable: true

And now a scenario like this one:

.. code-block::
  Feature: Test
  
    Scenario: Test
      Given I say Hello World

Will display our text as a pending text.

Conclusion
----------

Congratulations! You have just created a simple Behat extension.

> This extension demonstrates the three essential steps to building a Behat extension: defining an extension, creating an initializer, and configuring contexts.

Reviewer: "essential" isn't quite right here as a lot of extensions don't need to create initializers or configure contexts.

```suggestion
This extension demonstrates three of the common steps to building a Behat extension: defining an extension, creating an initializer, and configuring contexts.
```

Creating a Behat extension
==========================

Extensions are particularly useful when configuration becomes a necessity.

In this cookbook, we will create a simple extension named ``LogsExtension`` that logs scenario durations and cover:

#. Setting up the context
#. Creating the extension
#. Initializing the context

Setting Up the Context
----------------------

First, we need to create a class that will handle the logging. There are several ways to do this (you can hook
directly into behat's event dispatcher, for example).

For our case, we will create a ``Context`` and use the :doc:`normal tagged hooks</user_guide/context/hooks>`
that you might have used in your own contexts.

This is a straightforward approach, and means we can also show how to initialise any context with custom
configuration or dependencies.
The ``LogsExtension`` will provide a ``LogsContext`` that hooks into scenarios to log their start and end times.

Directory structure:

.. code-block::
  
  src/
      Context/
          LogsContext.php   # This is where we'll implement our logging logic

The code for ``LogsContext.php``:

.. code-block:: php

  <?php
  

>   namespace Behat\LogsExtension\Context;

Reviewer 2: I suggest using a different namespace that `Behat`, to avoid suggesting that extensions should be using the `Behat` namespace. The `Behat` namespace should be used only by official extensions. Third-party extensions should be using the top-level namespace of that third-party instead.

Author: Surely a typo, it's not in my own [boilerplate](https://github.com/monitaurus/behat-extension-boilerplate) 🤷 

It's fixed thanks: 8e8526805ae58c80b79ef17954b1c7dd6a9588be

While there might be alternatives, we will stick to the straightforward method.

The ``getConfigKey`` method is used to identify our extension in the configuration, and the ``configure`` method is used to define the configuration tree.

The code for ``LogsExtension.php``:

.. code-block:: php
  
  <?php
  
  namespace Behat\LogsExtension\ServiceContainer;
  
  use Behat\Testwork\ServiceContainer\Extension;
  use Behat\Testwork\ServiceContainer\ExtensionManager;
  use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
  use Symfony\Component\DependencyInjection\ContainerBuilder;
  
  class LogsExtension implements Extension
  {
      public function getConfigKey()
      {
          return 'logs_extension';
      }
  
      public function initialize(ExtensionManager $extensionManager)
      {
          // Empty for our case, but useful to hook into other extensions' configurations
      }

      public function configure(ArrayNodeDefinition $builder)
      {
          $builder
              ->addDefaultsIfNotSet()
              ->children()

>                   ->scalarNode('enable')->defaultFalse()->end()

Reviewer 2: this looks like a use case for a `booleanNode` instead

Author: Sure! Fixed: 8e8526805ae58c80b79ef17954b1c7dd6a9588be

To complete the extension, we must add methods to ``LogsContext`` to receive the configuration values and use those in the hooks:

.. code-block:: php

  // ...
  
  class LogsContext implements Context
  {
      private bool $enable = false;
      private string $filepath;
  
      public function initializeConfig(bool $enable, string $filepath)
      {
          $this->enable = $enable;
          $this->filepath = $filepath;
      }
  
      /** @BeforeScenario */
      public function before(BeforeScenarioScope $scope)
      {
          if ($this->enable === false) {
              return;
          }

          file_put_contents(
              $this->filepath,
              'START: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
              FILE_APPEND
          );
      }
  
      /** @AfterScenario */
      public function after(AfterScenarioScope $scope)
      {
          if ($this->enable === false) {
              return;
          }
  
          file_put_contents(
              $this->filepath,
              'END: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
              FILE_APPEND
          );
      }
  }

Conclusion
----------

> Congratulations! You have just created a simple Behat extension that logs scenario durations. This extension demonstrates the three essential steps to building a Behat extension: defining an extension, creating an initializer, and configuring contexts.

Reviewer 2: Using a context to log things is probably a bad example. It won't work out of the box, as your context needs to be registered in the config of the project to be used (which is not even described in your cookbook).
The extension should use events instead of hooks instead.

Author: > The extension should use events instead of hooks instead.

There's another way to access events apart hooks?
Have you got an example or a link to share?

Btw, I just push a new commit with an example of a `behat.yaml` configuration: f91f7c9

Reviewer: @monitaurus yes - you can instead [register an event subscriber](https://github.com/ingenerator/behat-support/blob/66028a7f3ed4195a3c042bc8e49966a197753144/src/Extension/SaveFailingPagesExtension.php#L49-L58) and [implement the EventSubscriberInterface](https://github.com/ingenerator/behat-support/blob/1.x/src/Extension/SaveFailingPagesExtension/SaveFailingPagesListener.php) - I'm not saying those are good examples overall BTW, I just knew where to find them quickly :laughing: 

There are a few advantages, partly your behat.yml doesn't have to register a context as well as an extension, partly that way you can set the priority of your subscriber so you can control when it runs relative to hooks and other listeners.

**HOWEVER** since this is a cookbook I would also be OK with keeping your example how it is, now that it mentions that there are other ways to do this and has a behat.yml showing how it works.

Your example has the advantage that:

* It also shows how an extension can initialise a context.
* It minimises the number of "new things" people need to consider. They have probably already written hooks in their own contexts and understand how they work. For example, the event listeners get slightly different payloads, and with the power to set priority also comes the ability to cause problems that don't exist with hooks.

And it works, so long as you put the right config in behat.yml.

So my temptation is to say as long as it's clear we're saying this is a simplified example, and not the only / best way (which I think it does) then this is useful advice for people.

Reviewer 2: A cookbook showing how an extension can initialize a context is indeed worth it. But then I would suggest 2 changes:

- not naming it `Creating a Behat extension`, as this cookbook is about initializing a context from an extension, which is only a specific thing an extension can do
- maybe find a better example for what the context is doing (it should probably be a context providing some step definition, to be a case where this is indeed the right tool to solve the need)

Author: _Oh my 😱 I didn't knew that.
At work I made a OpenAPI extension that I need to load the context every time.
I'll change it to use subscriber._

I agree @stof, for a log extension, a subscriber is far better.
But as @acoulton say, as it's a cookbook, I think we can leave it this way.
I also fear it may overload the text with another notion.

I suggest a direct follow up to this cookbook, a _"Subscribing to events"_ one.
It could use the same boilerplate as this cookbook, and make the step further to change hooks to subscribers.

Author: Ok, @stof @acoulton, I change the extension example to a simple Hello World.
The extension give a Context which will throw a PendingException which message is configurable in parameters.

The extension is clearly a nonsense, but allow to simplify a lot the whole cookbook by removing the hook.

What do you think?
